src/lc3/vm.rs

```diff
@@ -100,27 +107,179 @@ impl VM {
+            let res = value_in_r1 & imm5;
+            self.set_register(r0, res)?;
+        } else {
+            let r2 = instr & 0b0111;
+            let value_in_r2 = *self.get_register(r2)?;
+            let res = value_in_r1 & value_in_r2;
+            self.set_register(r0, res)?;
+        }
+        self.update_flags(r0)?;
+        Ok(())
+    }
+
+    fn not(&mut self, instr: u16) -> Result<(), VMError> {
+        let r0 = (instr >> 9) & 0b0111;
+        let r1 = (instr >> 6) & 0b0111;
+        let value_in_r1 = *self.get_register(r1)?;
+        self.set_register(r0, !value_in_r1)?;
+        self.update_flags(r0)?;
+        Ok(())
+    }
+
+    fn br(&mut self, instr: u16) -> Result<(), VMError> {
+        let pc_offset = sign_extend(instr & 0b0001_1111_1111, 9)?;
+        let cond_flag_instr = (instr >> 9) & 0b0111;
+        let meet_condition = match self.cond {
+            ConditionFlag::Neg => cond_flag_instr & 0b100,
+            ConditionFlag::Zro => cond_flag_instr & 0b010,
+            ConditionFlag::Pos => cond_flag_instr & 0b001,
+        };
+        if meet_condition != 0 {
+            self.pc = self
+                .pc
+                .checked_add(pc_offset)
+                .ok_or(VMError::Adding(String::from("Overflow in branching")))?;
+        }
+        Ok(())
+    }
+
+    fn jmp(&mut self, instr: u16) -> Result<(), VMError> {
+        let r1 = (instr >> 6) & 0b0111;
+        let value_in_r1 = *self.get_register(r1)?;
+        self.pc = value_in_r1;
+        Ok(())
+    }
+
+    fn jsr(&mut self, instr: u16) -> Result<(), VMError> {
+        let long_flag = (instr >> 11) & 1;
+        self.set_register(7, self.pc)?;
+        if long_flag == 1 {
+            let long_pc_offset = sign_extend(instr & 0b0111_1111_1111, 11)?;
+            self.pc = self
+                .pc
+                .checked_add(long_pc_offset)
+                .ok_or(VMError::Adding(String::from(
+                    "Overflow in jumping register",
+                )))?;
+        } else {
+            let r1 = (instr >> 6) & 0b0111;
+            let value_in_r1 = *self.get_register(r1)?;
+            self.pc = value_in_r1;
+        }
+        Ok(())
+    }
+    fn lea(&mut self, instr: u16) -> Result<(), VMError> {
+        let r0 = (instr >> 9) & 0b111;
+        let pc_offset = sign_extend(instr & 0b0001_1111_1111, 9)?;
+        let value = self
+            .pc
+            .checked_add(pc_offset)
+            .ok_or(VMError::Adding(String::from(
+                "Overflow in LEA offset addition",
+            )))?;
+        self.set_register(r0, value)?;
+        self.update_flags(r0)?;
+        Ok(())
+    }
+}
+
+fn sign_extend(mut value: u16, bit_count: u16) -> Result<u16, VMError> {
+    let last_bit_position = bit_count
+        .checked_sub(1)
+        .ok_or(VMError::Adding(String::from("Invalid last position bit")))?;
+    if (value >> (last_bit_position) & 1) == 1 {
+        value |= 0xFFFF << bit_count;
+    }
+    Ok(value)
 }
```

Comment:
It could be nice to have doc comments for the ISA.

Reply:
Perfect! I will be adding those